### DISCUSSION TITLE
built a generic cache solution

### DISCUSSION THREAD

**Discussion by earonesty** [sqlalchemy/sqlalchemy#11127] [Usage Questions] [answered]
so i built a nice generic cache solution that works for certain kinds of queries.   any query that joins one or more tables that have primary keys and doesn't aggregate is eligible.   the engine checks the cache for matching results and if none, ensures the primary keys are returned in the results, retrieves it, stuffs it in the cache, and then returns a proxy result object instead.   it can then subscribe to table updates using various techniques for invalidation support, and can use different cache engine backends.

problems:

 - proxy result object is far from identical to a regular Result object, which is extremely complex and is tied to the backend in ways so as not to be reusable
 - rollbar's technique of hooking the Session object and changing the query class is invasive, requiring the caller to have to manage their sessions differently, and can be extremely complex.
 
 ideas:
 
  - just keep building out the proxy?(
  - sneakily create a new session based on the passed in session that uses a CachedQuery 
  
  the interface i have is:
  
  cache.execute(session, query)
  
  this allows the caller to simple replace: session.execute(), which is a nice change.
  
  in many codebases, session handling is abstracted very far from the caller, but cache decisions rarely are made at the session level
  
  kinda not sure which way to go, looking for ideas

**Comment by zzzeek** [comment, marked as answer]
we have a comprehensive caching example that you can use the build off from.  the Result object is available to you in several forms including IteratorResult which you can pack with your own data.  however the caching example takes advantage of being able to take the existing Result and "freeze" it, allowing it to be reconstructed later.

Go through https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.dogpile_caching which should illustrate lots of interesting ways to do things and still have it look completely like the normal ORM examples.

**Reply by earonesty** [reply]
i see.   because of the structure, if a cache isn't "installed" correctly, it will silently ignore the user's request and there's no way for the FromCache() option to fail, and the user will never know that the cache is not being used.    would be nicer if hooking queries and returning results  was more of a "first class" feature (derive from query, or some other imperative operation), imo, and less tied to an event/hook system.

**Reply by zzzeek** [reply]
well that's waht I got, take a look to see how you can use the Result object with your own solution (knowing the internals is not necessary).

**Reply by earonesty** [reply]
im following that example, but not too closely.   definitely hooking the top orm execute is what's needed.   dp_cache is a bit odd (no async, remote locking can't be turned off if not needed, etc.) .    the silent failure thing is a beast to deal with in tests, etc.
